rust-symcrypt/src/ecc/mod.rs

```diff
@@ -0,0 +1,627 @@
+    // TODO: Discuss send/sync for rustls
+}
+
+unsafe impl Sync for InnerEcCurve {
+    // TODO: Discuss send/sync for rustls
+}
+
+// Must drop EcCurve after EcKey is dropped, will always be the case since EcCurve is static.
+impl Drop for InnerEcCurve {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            if self.0 != null_mut() {
+                symcrypt_sys::SymCryptEcurveFree(self.0);
+            }
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+/// `EcKeyUsage` will indicate if the [`EcKey`] will be used for [`EcKeyUsage::EcDsa`] or [`EcKeyUsage::EcDh`], or [`EcKeyUsage::EcDhAndEcDsa`].
+pub enum EcKeyUsage {
+    /// When using [`EcKeyUsage::EcDsa`] the intended usage of the key will be only for EcDsa sign or verify.
+    EcDsa,
+
+    /// When using [`EcKeyUsage::EcDh`] the intended usage of the key will be only for EcDh.
+    EcDh,
+
+    /// When using [`EcKeyUsage::EcDhAndEcDsa`] the intended usage of the key will be for both EcDsa and EcDh.
+    EcDhAndEcDsa,
+}
+
+// to_symcrypt_flag converts the EcKeyUsage enum to the corresponding SymCrypt flag and is only needed internally.
+impl EcKeyUsage {
+    pub(crate) fn to_symcrypt_flag(&self) -> u32 {
+        match self {
+            EcKeyUsage::EcDsa => symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDSA,
+            EcKeyUsage::EcDh => symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH,
+            EcKeyUsage::EcDhAndEcDsa => {
+                symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDSA | symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+/// ECC Key State Object.
+///
+/// `EcKey` is a struct that represents an Ec Key. It can hold either just an Ec public key, or an Ec key pair.
+/// To check if if the key has a private key attached, you can use [`EcKey::has_private_key()`]
+pub struct EcKey {
+    inner_key: InnerEcKey,
+    curve_type: CurveType,
+    ec_key_usage: EcKeyUsage,
+    has_private_key: bool,
+}
+
+unsafe impl Send for EcKey {
+    // TODO: Discuss send/sync for rustls
+}
+
+unsafe impl Sync for EcKey {
+    // TODO: Discuss send/sync for rustls
+}
+
+impl EcKey {
+    /// `generate_key_pair()` generates a random Ec Key object that has a public and private key attached based on the [`CurveType`] and [`EcKeyUsage`] provided.
+    ///
+    /// `curve_type` represents the [`CurveType`] of the key.
+    ///
+    /// `ec_key_usage` represents the [`EcKeyUsage`] of the key and indicates if the key will be used for [`EcKeyUsage::EcDsa`] or [`EcKeyUsage::EcDh`], or [`EcKeyUsage::EcDhAndEcDsa`].
+    ///
+    /// This function will return a [`SymCryptError::MemoryAllocationFailure`] if there is not enough memory to allocate the key.
+    pub fn generate_key_pair(
+        curve_type: CurveType,
+        ec_key_usage: EcKeyUsage,
+    ) -> Result<Self, SymCryptError> {
+        let ec_curve = InnerEcCurve::new(curve_type); // Can fail here due to insufficient memory.
```

Comment:
>// Can fail here due to insufficient memory. [](http://example.com/codeflow?start=54&length=44)

Need to check this for null before calling `SymCryptEckeyAllocate`, don't you?

Reply:
at this point `ec_curve` is not null since we've done the check in `internal_new()`

To my knowledge, memory allocation error would only ever occur in this case if the entire system ran out of memory, which would cause Rust to crash anyways 


```
// Stack allocated since SymCryptEcCurveAllocate is called.
let curve_ptr = symcrypt_sys::SymCryptEcurveAllocate(to_symcrypt_curve(curve), 0);
if curve_ptr.is_null() {
  return Err(SymCryptError::MemoryAllocationFailure);
}
// Curve needs to be wrapped to properly free the curve in the case there is an error in future initialization in EcDsa or EcDh.
Ok(InnerEcCurve(curve_ptr))

```

Reply:
That is true of pretty much all memory allocation failures - they're rare on modern systems with large amounts of RAM and virtual memory. But it's still good to handle them explicitly so that you know what the behavior is in the error case.

If the expected behavior is for the program to panic when it can't allocate these curves (which seems sensible given that they're "static" objects with respect to the lifetime of the program), I would recommend doing that explicitly in `InnerEcCurve::new`

Reply:
Cleared up via Teams chat - I misunderstood how this worked in Rust. The current way you're doing it is fine.

I do think removing these comments would be good; they just lead to confusion.